#' Locate Clustered SNPs in sequence
#'
#' @param SeqAligned list of sequence vectors \code{\link{alignment2Fasta}}
#' @param ClusterTable Table generated using getClusterTable function \code{\link{getClusterTable}}
#' @return List of sequences , where clustered SNPs are marked by number
SeqLocateCluster<-function(SeqAligned,ClusterTable)
{
  #Converting alignment to list
  SeqList<-list()
  for (seq in 1:length(SeqAligned)) {
    SeqAlignedOnePiece<-paste(SeqAligned[[seq]], collapse = '')
    SeqList[[seq]]<-SeqAlignedOnePiece
  }
  # Locate clusters SNPs and replace it with designated number
  SeqListArrays<-list()
  #Its programmatically less processing effort to split first
  for (SeqN in 1:length(SeqList)) {
    Seq<-SeqList[[SeqN]]
    SeqArray<-strsplit(Seq,"")
    SeqListArrays[SeqN]<-SeqArray
    #Do through the loop
    for (SNP in 1:nrow(ClusterTable)) {
      pos<-as.numeric(gsub("N","",ClusterTable[SNP,]$Nucleotide))
      clu<-ClusterTable[SNP,]$Cluster
      if (SeqListArrays[[SeqN]][pos] != "-") {
        SeqListArrays[[SeqN]][pos] <- clu
      }
    }
  }
  SeqListArrays
}



#' Get Cluster annotation table for SNP groups in sequences
#'
#' @param SeqInfo  Information of the sequences \code{\link{getSeqInfo}}
#' @param SeqWithClusters Sequences with clusters generated by \code{\link{SeqLocateCluster}}
#' @param ClusterMax Max number of clusters in the sequences
#' @param MinimumClusterLength Minimum length of clustered nucleotides to consider as a group cluster
#' @return Table
#' @importFrom  stringr str_locate_all
#' @import tidyverse
getClusterAnnotation<-function(SeqInfo,SeqWithClusters,ClusterMax=3,MinimumClusterLength=3)
{
  #Table Structure
  clusterAnnoTable<-data.frame(molecule="",gene="",start=0,end=0,
                               strand="",subgene="",from=0,to=0,orientation=0)
  #Loop in sequences
  for (seqn in 1:length(SeqWithClusters)) {
    GeneSeq<-paste0(SeqWithClusters[[seqn]],collapse = "")
    GeneSeq<-gsub("-","",GeneSeq)
    #Loop in Clsuters
    for (clu in 1:ClusterMax) {
      locations<-str_locate_all(GeneSeq, paste0("[\\",clu,"]+"))
      locations<-as.data.frame(locations)
      for (i in 1:nrow(locations)) {
        matchstart<-locations[i,]$start
        matchend<-locations[i,]$end
        Found<-substr(GeneSeq,matchstart,matchend)
        if (nchar(Found)>0 && (matchend-matchstart>=MinimumClusterLength)) {
          clusterAnnoTable<-clusterAnnoTable %>% add_row(
            molecule=SeqInfo[seqn,]$Name, #"molecule	gene"
            gene=paste0("Cluster",clu), #Cluster Name
            start=0,                   #gene start from 0 default
            end=SeqInfo[seqn,]$Length, #gene Length/end
            strand="forward",             #strand forward default
            subgene=paste0("Cluster",clu), #Cluster Name
            from=matchstart,          #SNP group cluster position start
            to=matchend,            #SNP group cluster position end
            orientation=1)                   #orientation default 1
        }
      }
    }
  }
  clusterAnnoTable[-1,]
}

#' Return Gene information table for clusterplot
#' @param SeqInfo Information of the sequences \code{\link{getSeqInfo}}
#' @return data.frame
GeneInfoForClusterPlot<-function(SeqInfo)
{
  GeneInfoTable<-as.data.frame(matrix(nrow = nrow(SeqInfo), ncol = 6))
  colnames(GeneInfoTable)<-c("molecule","gene","start","end","strand","orientation")
  GeneInfoTable$molecule<-SeqInfo$Name
  GeneInfoTable$gene<-SeqInfo$Name
  GeneInfoTable$start<-0
  GeneInfoTable$end<-SeqInfo$Length
  GeneInfoTable$strand<-"forward"
  GeneInfoTable$orientation<-1
  GeneInfoTable
}

#' Plot SNP cluster on genes along with Phylogenetic tree
#'
#' @param SeqInfo table of sequence information generated using \code{\link{getSeqInfo}}
#' @param Alignment sequence alignment object
#' @param MaxCluster Maximum number of clusters to draw default 3
#' @param MinimumClusterLength Minimum number of base pairs for a cluster to be consider 3bp.
#' @param Cluster object of HCPC \code{\link{SNPCluster}}
#'
#' @import ggplot2
#' @import gggenes
#' @import ggtree
#' @import forcats
#' @import cowplot
#' @return plot
#' @export
SNPClusterPlotWithTree<-function(SeqInfo,Alignment,Cluster,MaxCluster=3,MinimumClusterLength=3)
{
  DistanceTable<-getDistanceMatrixTabel(SeqInfo,Alignment)
  ClustTable<-getClusterTable(Cluster)
  SeqAligned<- alignment2Fasta(Alignment,SeqInfo)
  SeqWithClusters<-SeqLocateCluster(SeqAligned,ClustTable)
  SNPClusterAnnotation<-getClusterAnnotation(SeqInfo,SeqWithClusters,MaxCluster,MinimumClusterLength)
  GeneInfoTable<-GeneInfoForClusterPlot(SeqInfo)

  ##### Tree
  options(ignore.negative.edge=TRUE)
  myTree<-getTree(DistanceTable)
  # SeqTreePlot<-ggplot(myTree, aes(x, y)) + geom_tree() + theme_tree()+
  #   theme(plot.margin=unit(c(0,0,0,0), "cm"))
  SeqTreePlot <- ggtree(myTree) + geom_tiplab(align=TRUE,size=3) + hexpand(.4)

  #Tree order
  TreeTable=fortify(myTree)
  TipTable = subset(TreeTable, isTip)
  NodesOrder<-TipTable$label[order(TipTable$y, decreasing=TRUE)]


  GeneInfoTable<-GeneInfoTable[match(rev(NodesOrder), GeneInfoTable$molecule),]
  ##### Sequences
  SeqClPlot<-GeneInfoTable %>% ggplot(aes(xmin = start, xmax = end, y = molecule))+
    geom_gene_arrow(fill = "white")+ aes(y = fct_inorder(gene))+
    geom_subgene_arrow(data = SNPClusterAnnotation,
                        aes(xmin = start, xmax = end,
                            y = molecule, fill = gene,
                            xsubmin = from, xsubmax = to),
                        color="black", alpha=.7) + theme_genes()+
    guides(fill=guide_legend(title="SNP Clusters"))+
    theme(axis.title.y = element_blank()) +
    theme(plot.margin=unit(c(0,0,0,0), "cm"))+
    #### Prepare part
    theme_tree2() +
    xlab(NULL) + ylab(NULL)+
    theme_minimal()+ theme(
      panel.grid = element_blank(),
      axis.title = element_blank(),
      axis.text = element_blank(),
      axis.ticks.x = element_blank())

  plot_grid(SeqTreePlot,SeqClPlot, ncol=2,rel_widths = c(1,2),label_size = 1)

}


#' Plot SNP cluster locations on genes
#' @param SeqInfo table of sequence information generated using \code{\link{getSeqInfo}}
#' @param SeqAligned list of aligned vectors contains aligned sequences using \code{\link{alignment2Fasta}}
#' @param Cluster object of HCPC \code{\link{SNPCluster}}
#' @param MaxCluster Maximum number of clusters to draw default 3
#' @param MinimumClusterLength Minimum number of base pairs for a cluster to be consider 3bp.
#' @return plot
#' @export
#' @import ggplot2
#' @import gggenes
SNPClusterPlot<-function(SeqInfo,SeqAligned,Cluster,MaxCluster,MinimumClusterLength)
{

  ClustTable<-getClusterTable(Cluster)
  SeqWithClusters<-SeqLocateCluster(SeqAligned,ClustTable)
  SNPClusterAnnotation<-getClusterAnnotation(SeqInfo,SeqWithClusters,MaxCluster,MinimumClusterLength)
  GeneInfoTable<-GeneInfoForClusterPlot(SeqInfo)

  SeqClPlot<- ggplot(GeneInfoTable,aes(xmin = start, xmax = end, y = molecule))+
    geom_gene_arrow(fill = "white")+ aes(y = fct_inorder(gene))+
    geom_subgene_arrow(data = SNPClusterAnnotation,
                       aes(xmin = start, xmax = end,
                           y = molecule, fill = gene,
                           xsubmin = from, xsubmax = to),
                       color="black", alpha=.7) + theme_genes()+
    guides(fill=guide_legend(title="SNP Clusters"))+
    theme(axis.title.y = element_blank()) +
    theme(plot.margin=unit(c(0,0,0,0), "cm"))

  SeqClPlot

}

